workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,228 @@
         {
-            // Mark readers as to be enabled
+            // Mark reader components ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = behavioursRequiringReaderTypes[componentId];
+            MarkComponentsReady(relevantReaderSpatialOSBehaviours);
         }
 
         public void RemoveComponent(uint componentId)
         {
-            // Mark readers as to be disabled
+            // Mark reader components not ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = behavioursRequiringReaderTypes[componentId];
+            MarkComponentsNotReady(relevantReaderSpatialOSBehaviours);
         }
 
         public void ChangeAuthority(uint componentId, Authority authority)
         {
-            // Mark writers as to be enabled/disabled
+            if (authority == Authority.Authoritative)
+            {
+                // Mark writer components ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = behavioursRequiringWriterTypes[componentId];
+                MarkComponentsReady(relevantWriterSpatialOSBehaviours);
+            }
+            else if (authority == Authority.NotAuthoritative)
+            {
+                // Mark writer components not ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = behavioursRequiringWriterTypes[componentId];
+                MarkComponentsNotReady(relevantWriterSpatialOSBehaviours);
+            }
+        }
+
+        private void MarkComponentsReady(IEnumerable<MonoBehaviour> behaviours)
+        {
+            // Inject all Readers/Writers at once when all requirements are met
+            foreach (var behaviour in behaviours)
+            {
+                numUnsatisfiedReadersOrWriters[behaviour]--;
+                if (numUnsatisfiedReadersOrWriters[behaviour] == 0)
+                {
+                    // Schedule activation
+                    if (!behaviour.enabled)
+                    {
+                        behavioursToEnable.Add(behaviour);
+                    }
+                    else
+                    {
+                        behavioursToDisable.Remove(behaviour);
+                    }
+                }
+            }
+        }
+
+        private void MarkComponentsNotReady(IEnumerable<MonoBehaviour> behaviours)
+        {
+            foreach (var behaviour in behaviours)
+            {
+                // De-inject all Readers/Writers at once when a single requirement is not met
+                if (numUnsatisfiedReadersOrWriters[behaviour] == 0)
+                {
+                    // Schedule deactivation
+                    if (behaviour.enabled)
+                    {
+                        behavioursToDisable.Add(behaviour);
+                    }
+                    else
+                    {
+                        behavioursToEnable.Remove(behaviour);
```

Comment:
Are we guaranteed that `behavioursToEnable` will contain `behaviour` ?

Reply:
The code assumes that no other system interferes with enabling/disabling these MonoBehaviours, so if something has all Readers/Writers ready but is not enabled, it must have been marked for enabling in this tick. That is something I am allowed to assume, right?